File: src/lib/git.ts
import { invoke } from "@tauri-apps/api/core";

export type GitFileStatus = {
  path: string;
  status: string;
};

export type GitStatusSnapshot = {
  staged: GitFileStatus[];
  unstaged: GitFileStatus[];
  /** Files with unresolved merge conflicts (status "U"). Only populated while mergeInProgress is true. */
  conflicted: GitFileStatus[];
  branch: string | null;
  /** Whether HEAD resolves to a commit (false on a brand-new, empty repo). */
  hasCommits: boolean;
  /** Whether the current branch has an upstream remote-tracking branch configured. */
  hasUpstream: boolean;
  /** Commits on HEAD not yet on the upstream (local-only; 0 when hasUpstream is false). */
  ahead: number;
  /** Whether a merge was left unfinished by a conflict (MERGE_HEAD present). */
  mergeInProgress: boolean;
};

export type GitConflictFile = {
  path: string;
  content: string;
};

export function hasUnpushedCommits(status: GitStatusSnapshot): boolean {
  if (!status.hasCommits) return false;
  return !status.hasUpstream || status.ahead > 0;
}

export type GitCommitSummary = {
  hash: string;
  message: string;
  author: string;
  time: number;
};

export type PullMode = "merge" | "rebase";

export type GitDiffScope = "staged" | "unstaged";

export type GitFileDiff = {
  original: string;
  modified: string;
  originalLabel: string;
  modifiedLabel: string;
  isBinary: boolean;
};

export type GitBranchInfo = {
  name: string;
  isCurrent: boolean;
  isRemote: boolean;
  /** Commits on the upstream not yet pulled locally; null if no upstream. */
  behind: number | null;
  /** The branch's tip commit oid — used to undo a branch deletion. */
  tipOid: string | null;
};

export function hasTrackedGitChanges(status: GitStatusSnapshot): boolean {
  return (
    status.staged.length > 0 ||
    status.unstaged.some((file) => file.status !== "?")
  );
}

/** Coarse-grained summary of repo sync state for the TopBar pill. Distinct
 * from GitSyncStatus (the live {ahead,behind} fetch-check shape below) —
 * this is a UI-facing enum derived from already-available snapshot data,
 * recomputed on the same cadence git.status/branches already refresh on
 * (no new polling). */
export type SyncPillState =
  | "conflict"
  | "merging"
  | "dirty"
  | "behind"
  | "unpushed"
  | "synced";

export type SyncPillInput = {
  status: GitStatusSnapshot;
  /** `behind` of the *current* branch from GitBranchInfo — a last-fetch snapshot, not live. */
  currentBranchBehind: number | null;
  /** Whether a stash-restore conflict is pending (App.tsx's pendingStashConflict). */
  hasPendingStashConflict: boolean;
};

/** Precedence: conflict > merging > dirty > behind > unpushed > synced.
 * `dirty` outranks `behind`/`unpushed` because it's the immediate local
 * loss-risk; `behind` outranks `unpushed` because pushing while behind
 * needs a pull first anyway — telling the user to pull is the more useful
 * of the two messages. */
export function deriveSyncPillState(input: SyncPillInput): SyncPillState {
  const { status, currentBranchBehind, hasPendingStashConflict } = input;
  if (status.conflicted.length > 0 || hasPendingStashConflict) return "conflict";
  if (status.mergeInProgress) return "merging";
  if (hasTrackedGitChanges(status)) return "dirty";
  if ((currentBranchBehind ?? 0) > 0) return "behind";
  if (hasUnpushedCommits(status)) return "unpushed";
  return "synced";
}

export function syncPillLabel(state: SyncPillState): string {
  switch (state) {
    case "conflict":
      return "Конфликт файлов";
    case "merging":
      return "Слияние в процессе";
    case "dirty":
      return "Есть несохранённые правки";
    case "behind":
      return "Есть новые изменения";
    case "unpushed":
      return "Нужно отправить";
    case "synced":
      return "Все синхронизировано";
  }
}

/** Compact label for the TopBar sync pill — full text stays in `title`. */
export function syncPillShortLabel(state: SyncPillState): string {
  switch (state) {
    case "conflict":
      return "Конфликт";
    case "merging":
      return "Слияние";
    case "dirty":
      return "Изменения";
    case "behind":
      return "Обновить";
    case "unpushed":
      return "Отправить";
    case "synced":
      return "Синхронизировано";
  }
}

export type GitSyncStatus = {
  ahead: number;
  behind: number;
};

/** One shelved (auto-stashed) set of tracked working-tree changes. */
export type GitStashEntry = {
  id: string;
  branch: string;
  createdAt: number;
  filesChanged: number;
};

export type GitStashRestoreOutcome =
  | { outcome: "applied"; entry: GitStashEntry }
  | { outcome: "conflict"; entry: GitStashEntry }
  | { outcome: "blocked"; entry: GitStashEntry; reason: string }
  | { outcome: "skipped"; count: number };

export type CheckoutOutcome = {
  shelved: GitStashEntry | null;
  restore: GitStashRestoreOutcome | null;
};

export type GitProgressEvent =
  | { kind: "started"; op: string }
  | {
      kind: "transfer";
      op: string;
      receivedObjects: number;
      totalObjects: number;
      receivedBytes: number;
      indexedDeltas: number;
      totalDeltas: number;
    }
  | { kind: "push"; op: string; current: number; total: number; bytes: number }
  | { kind: "finished"; op: string };

export const GIT_PROGRESS_EVENT = "git://progress";

export type SshKeySource =
  | { kind: "keyContent"; privateKey: string }
  | { kind: "keyFile"; path: string };

export type SshKeyConfig = {
  name: string;
  host?: string;
  source: SshKeySource;
  passphrase?: string;
};

export type GitCredentials = {
  sshKeys: SshKeyConfig[];
  trustAllSshHostKeys: boolean;
};

export type AppKeyStatus = {
  exists: boolean;
  publicKey: string;
  privateKeyAvailable: boolean;
  isImported: boolean;
};

export function gitStatus(repoRoot: string): Promise<GitStatusSnapshot> {
  return invoke<GitStatusSnapshot>("git_status", { repoRoot });
}

export function gitStage(repoRoot: string, paths: string[]): Promise<void> {
  return invoke<void>("git_stage", { repoRoot, paths });
}

export function gitUnstage(repoRoot: string, paths: string[]): Promise<void> {
  return invoke<void>("git_unstage", { repoRoot, paths });
}

export function gitCommit(
  repoRoot: string,
  message: string,
): Promise<string> {
  return invoke<string>("git_commit", { repoRoot, message });
}

export function gitLog(
  repoRoot: string,
  limit = 20,
): Promise<GitCommitSummary[]> {
  return invoke<GitCommitSummary[]>("git_log", { repoRoot, limit });
}

export function gitUnpushedCommits(
  repoRoot: string,
  limit = 50,
): Promise<GitCommitSummary[]> {
  return invoke<GitCommitSummary[]>("git_unpushed_commits", { repoRoot, limit });
}

export function gitIncomingCommits(
  repoRoot: string,
  limit = 50,
): Promise<GitCommitSummary[]> {
  return invoke<GitCommitSummary[]>("git_incoming_commits", { repoRoot, limit });
}

export function gitPull(repoRoot: string, mode: PullMode): Promise<void> {
  return invoke<void>("git_pull", { repoRoot, mode });
}

export function gitConflictFileContent(
  repoRoot: string,
  path: string,
): Promise<GitConflictFile> {
  return invoke<GitConflictFile>("git_conflict_file_content", { repoRoot, path });
}

export function gitResolveConflict(
  repoRoot: string,
  path: string,
  content: string,
): Promise<void> {
  return invoke<void>("git_resolve_conflict", { repoRoot, path, content });
}

export function gitFinishMerge(repoRoot: string): Promise<string> {
  return invoke<string>("git_finish_merge", { repoRoot });
}

export function gitAbortMerge(repoRoot: string): Promise<void> {
  return invoke<void>("git_abort_merge", { repoRoot });
}

export function gitResetToRemote(repoRoot: string): Promise<void> {
  return invoke<void>("git_reset_to_remote", { repoRoot });
}

export function gitSyncStatus(repoRoot: string): Promise<GitSyncStatus> {
  return invoke<GitSyncStatus>("git_sync_status", { repoRoot });
}

export function gitPush(repoRoot: string): Promise<void> {
  return invoke<void>("git_push", { repoRoot });
}

export function gitFileDiff(
  repoRoot: string,
  path: string,
  scope: GitDiffScope,
): Promise<GitFileDiff> {
  return invoke<GitFileDiff>("git_file_diff", { repoRoot, path, scope });
}

export function gitCommitFiles(
  repoRoot: string,
  commitHash: string,
): Promise<GitFileStatus[]> {
  return invoke<GitFileStatus[]>("git_commit_files", { repoRoot, commitHash });
}

export function gitCommitFileDiff(
  repoRoot: string,
  commitHash: string,
  path: string,
): Promise<GitFileDiff> {
  return invoke<GitFileDiff>("git_commit_file_diff", {
    repoRoot,
    commitHash,
    path,
  });
}

/** Discards `path`'s uncommitted changes, taking a backup first so the
 * discard is undoable. Returns the backup id (opaque — pass to
 * gitRestoreDiscardBackup to undo) or null if there was nothing to
 * discard. */
export function gitDiscardFileChanges(
  repoRoot: string,
  path: string,
): Promise<string | null> {
  return invoke<string | null>("git_discard_file_changes", { repoRoot, path });
}

export function gitRestoreDiscardBackup(
  repoRoot: string,
  backupId: string,
): Promise<void> {
  return invoke<void>("git_restore_discard_backup", { repoRoot, backupId });
}

export function gitUndoCommit(repoRoot: string, commitHash: string): Promise<void> {
  return invoke<void>("git_undo_commit", { repoRoot, commitHash });
}

export function gitCreateBranchAtOid(
  repoRoot: string,
  name: string,
  oid: string,
): Promise<void> {
  return invoke<void>("git_create_branch_at_oid", { repoRoot, name, oid });
}

export function gitResetToOid(repoRoot: string, oid: string): Promise<void> {
  return invoke<void>("git_reset_to_oid", { repoRoot, oid });
}

export function gitHeadOid(repoRoot: string): Promise<string> {
  return invoke<string>("git_head_oid", { repoRoot });
}

export function gitApplyDiffContent(
  repoRoot: string,
  path: string,
  scope: GitDiffScope,
  content: string,
): Promise<void> {
  return invoke<void>("git_apply_diff_content", { repoRoot, path, scope, content });
}

export function gitListBranches(repoRoot: string): Promise<GitBranchInfo[]> {
  return invoke<GitBranchInfo[]>("git_list_branches", { repoRoot });
}

export function gitFetchBranches(repoRoot: string): Promise<void> {
  return invoke<void>("git_fetch_branches", { repoRoot });
}

export function gitCreateBranch(
  repoRoot: string,
  name: string,
  discardChanges = false,
): Promise<void> {
  return invoke<void>("git_create_branch", { repoRoot, name, discardChanges });
}

export function gitCheckoutBranch(
  repoRoot: string,
  name: string,
  discardChanges = false,
): Promise<CheckoutOutcome> {
  return invoke<CheckoutOutcome>("git_checkout_branch", { repoRoot, name, discardChanges });
}

export function gitDeleteBranch(
  repoRoot: string,
  name: string,
): Promise<void> {
  return invoke<void>("git_delete_branch", { repoRoot, name });
}

export function gitCheckoutRemoteBranch(
  repoRoot: string,
  name: string,
  discardChanges = false,
): Promise<CheckoutOutcome> {
  return invoke<CheckoutOutcome>("git_checkout_remote_branch", {
    repoRoot,
    name,
    discardChanges,
  });
}

export function gitStashList(repoRoot: string): Promise<GitStashEntry[]> {
  return invoke<GitStashEntry[]>("git_stash_list", { repoRoot });
}

export function gitStashApply(
  repoRoot: string,
  stashId: string,
): Promise<GitStashRestoreOutcome> {
  return invoke<GitStashRestoreOutcome>("git_stash_apply", { repoRoot, stashId });
}

export function gitStashDrop(repoRoot: string, stashId: string): Promise<void> {
  return invoke<void>("git_stash_drop", { repoRoot, stashId });
}

export function gitGetCredentials(): Promise<GitCredentials> {
  return invoke<GitCredentials>("git_get_credentials");
}

export function gitSaveCredentials(
  credentials: GitCredentials,
): Promise<void> {
  return invoke<void>("git_save_credentials", { credentials });
}

import { type ProbeResult } from "./project";

export type { ProbeResult };

export function gitClone(
  url: string,
  destination: string,
): Promise<ProbeResult> {
  return invoke<ProbeResult>("git_clone", { url, destination });
}

export function gitGetKeyStatus(): Promise<AppKeyStatus> {
  return invoke<AppKeyStatus>("git_get_key_status");
}

export function gitGenerateKey(): Promise<AppKeyStatus> {
  return invoke<AppKeyStatus>("git_generate_key");
}

export function gitImportKey(sourcePath: string): Promise<AppKeyStatus> {
  return invoke<AppKeyStatus>("git_import_key", { sourcePath });
}

File: src-tauri/src/domain/git.rs
use serde::{Deserialize, Serialize};
use thiserror::Error;

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitFileStatus {
    pub path: String,
    /// Single-letter status: M, A, D, R, ?
    pub status: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitStatusSnapshot {
    pub staged: Vec<GitFileStatus>,
    pub unstaged: Vec<GitFileStatus>,
    /// Files with unresolved merge conflicts (status "U"). Computed from the
    /// index's conflict stages directly — independent of `merge_in_progress`,
    /// since the index can hold conflicted entries even without a MERGE_HEAD
    /// (e.g. an interrupted merge, or state staged outside a `git merge`).
    pub conflicted: Vec<GitFileStatus>,
    pub branch: Option<String>,
    /// Whether HEAD resolves to a commit (false on a brand-new, empty repo).
    pub has_commits: bool,
    /// Whether the current branch has an upstream remote-tracking branch configured.
    pub has_upstream: bool,
    /// Commits on HEAD not yet on the upstream (purely local — reflects the
    /// last fetch, not live network state). Only meaningful when `has_upstream`
    /// is true; `0` otherwise.
    pub ahead: usize,
    /// Whether a merge was left unfinished by a conflict (MERGE_HEAD present).
    pub merge_in_progress: bool,
}

/// A conflicted file's current on-disk content, including conflict markers.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitConflictFile {
    pub path: String,
    pub content: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCommitSummary {
    pub hash: String,
    pub message: String,
    pub author: String,
    /// Unix timestamp (seconds).
    pub time: i64,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum PullMode {
    Merge,
    Rebase,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum GitDiffScope {
    Staged,
    Unstaged,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitFileDiff {
    pub original: String,
    pub modified: String,
    pub original_label: String,
    pub modified_label: String,
    pub is_binary: bool,
}

/// One contiguous run of lines attributed to the same commit — produced by
/// `infra::git_repo::blame` and surfaced to the AI harness as
/// `ToolResult::GitBlame`.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitBlameHunk {
    /// 1-indexed, inclusive.
    pub start_line: u32,
    /// 1-indexed, inclusive.
    pub end_line: u32,
    /// Short (7-char) commit hash.
    pub commit: String,
    pub author: String,
    /// ISO-8601 UTC (`YYYY-MM-DDTHH:MM:SSZ`).
    pub authored_at: String,
    pub summary: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitBranchInfo {
    pub name: String,
    pub is_current: bool,
    pub is_remote: bool,
    /// Commits present on the branch's upstream but not yet pulled locally.
    /// `None` when the branch has no upstream (or is itself a remote branch).
    pub behind: Option<usize>,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitSyncStatus {
    pub ahead: usize,
    pub behind: usize,
}

/// One shelved (auto-stashed) set of tracked working-tree changes, tied to
/// the branch it was captured from. Backed by a real git2 stash entry
/// tagged with the `docflow-auto: ` message prefix — hand-made `git stash`
/// entries created outside the app are filtered out and never surfaced or
/// touched by any of the operations below.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct GitStashEntry {
    /// Full stash commit oid (hex). Stable across index shifts, unlike
    /// libgit2's positional stash index, which shifts every time any stash
    /// is pushed or dropped.
    pub id: String,
    pub branch: String,
    /// Unix timestamp (seconds) — the stash commit's author time.
    pub created_at: i64,
    pub files_changed: usize,
}

/// Outcome of applying a shelf entry, either as part of an automatic
/// restore-on-checkout or a manual "Восстановить" action. Never implies the
/// entry was silently lost: `Conflict` and `Blocked` both leave the stash
/// entry in place so it stays visible and recoverable in the shelf list.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(tag = "outcome", rename_all = "camelCase")]
pub enum GitStashRestoreOutcome {
    /// Applied cleanly; the stash entry has already been dropped.
    Applied { entry: GitStashEntry },
    /// Conflict markers were written to the working tree and the index has
    /// conflicts (surfaced via the existing `GitStatusSnapshot.conflicted`).
    /// The stash entry was NOT dropped and remains in the shelf until the
    /// conflict is resolved.
    Conflict { entry: GitStashEntry },
    /// Restore was refused (e.g. the destination branch already has staged
    /// changes). Stash entry untouched.
    Blocked { entry: GitStashEntry, reason: String },
    /// More than one shelf entry exists for this branch — ambiguous, left
    /// for a manual pick via the shelf list instead of guessing which to
    /// apply.
    Skipped { count: usize },
}

/// Result of `checkout_branch`/`checkout_remote_branch` under the
/// auto-stash flow: tracked changes on the source branch are shelved
/// instead of blocking the switch, and any shelf entry for the destination
/// branch is auto-restored when unambiguous.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct CheckoutOutcome {
    /// Some(..) if tracked changes on the source branch were auto-stashed.
    pub shelved: Option<GitStashEntry>,
    /// Some(..) if an auto-restore was attempted (or skipped as ambiguous)
    /// for the destination branch.
    pub restore: Option<GitStashRestoreOutcome>,
}

/// A progress update for a network-bound git operation (fetch/pull/push/
/// clone), emitted as Tauri events so the UI can show real progress instead
/// of a static "in progress" label.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(tag = "kind", rename_all = "camelCase")]
pub enum GitProgressEvent {
    Started { op: String },
    Transfer {
        op: String,
        received_objects: usize,
        total_objects: usize,
        received_bytes: usize,
        indexed_deltas: usize,
        total_deltas: usize,
    },
    Push {
        op: String,
        current: usize,
        total: usize,
        bytes: usize,
    },
    Finished { op: String },
}

/// SSH key stored in app settings.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(tag = "kind")]
pub enum SshKeySource {
    /// Key content pasted directly into the app.
    #[serde(rename = "keyContent")]
    KeyContent {
        #[serde(rename = "privateKey")]
        private_key: String,
    },
    /// Path to a key file on disk.
    #[serde(rename = "keyFile")]
    KeyFile { path: String },
}

/// A named SSH key configuration, optionally scoped to a host.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SshKeyConfig {
    pub name: String,
    /// Optional host pattern (e.g. "bitbucket.company.com") for auto-selection.
    pub host: Option<String>,
    pub source: SshKeySource,
    /// Optional passphrase for encrypted keys.
    pub passphrase: Option<String>,
}

/// Collection of git credentials stored by the app.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GitCredentials {
    #[serde(default)]
    pub ssh_keys: Vec<SshKeyConfig>,
    /// When true, accept any SSH host key on first connection (trust-on-first-use).
    /// When false, rely on ~/.ssh/known_hosts for host verification.
    #[serde(default = "default_true")]
    pub trust_all_ssh_host_keys: bool,
}

fn default_true() -> bool {
    true
}

impl Default for GitCredentials {
    fn default() -> Self {
        Self {
            ssh_keys: Vec::new(),
            trust_all_ssh_host_keys: true,
        }
    }
}

/// Status of the app-managed SSH key.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AppKeyStatus {
    pub exists: bool,
    pub public_key: String,
    /// Whether the private key is available (decryptable).
    pub private_key_available: bool,
    /// Whether the user imported an external key instead of using the generated one.
    pub is_imported: bool,
}

/// App-managed key configuration stored in ~/.atlas/key_config.json.
#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct KeyConfig {
    pub public_key: String,
    /// Path to encrypted private key file, relative to settings dir.
    pub encrypted_private_key_path: String,
    /// True if the key was imported by the user (vs. generated by the app).
    pub is_imported: bool,
}

#[derive(Debug, Error)]
pub enum GitError {
    #[error("path is not a git repository: {0}")]
    NotARepository(String),
    #[error("failed to open repository: {0}")]
    Open(#[source] git2::Error),
    #[error("git operation failed: {0}")]
    Operation(#[source] git2::Error),
    #[error("commit message is empty")]
    EmptyMessage,
    #[error("nothing staged to commit")]
    NothingStaged,
    #[error(
        "git user.name / user.email are not configured; set them with git config user.name / user.email"
    )]
    MissingIdentity,
    #[error("invalid path: {0}")]
    InvalidPath(String),
    #[error("current branch has no upstream remote tracking branch")]
    NoUpstream,
    #[error("merge conflict; resolve conflicts manually and try again")]
    MergeConflict,
    #[error("rebase conflict; resolve conflicts manually and try again")]
    RebaseConflict,
    #[error("branch not found: {0}")]
    BranchNotFound(String),
    #[error("branch already exists: {0}")]
    BranchAlreadyExists(String),
    #[error("cannot delete the current branch")]
    CannotDeleteCurrentBranch,
    #[error("commit or discard tracked changes before switching branches")]
    CheckoutBlocked,
    #[error("clone failed: {0}")]
    CloneFailed(String),
    #[error("destination already exists: {0}")]
    DestinationExists(String),
    #[error("shelved changes not found: {0}")]
    StashNotFound(String),
    #[error("{0}")]
    Message(String),
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn ssh_key_source_key_content_serialization() {
        let source = SshKeySource::KeyContent {
            private_key: "test-key-data".into(),
        };
        let json = serde_json::to_string(&source).unwrap();
        assert!(json.contains(r#""kind":"keyContent""#));

        let parsed: SshKeySource = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed, source);
    }

    #[test]
    fn ssh_key_source_key_file_serialization() {
        let source = SshKeySource::KeyFile {
            path: "/home/user/.ssh/id_ed25519".into(),
        };
        let json = serde_json::to_string(&source).unwrap();
        assert_eq!(json, r#"{"kind":"keyFile","path":"/home/user/.ssh/id_ed25519"}"#);

        let parsed: SshKeySource = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed, source);
    }

    #[test]
    fn ssh_key_config_full_serialization() {
        let config = SshKeyConfig {
            name: "Bitbucket Work".into(),
            host: Some("bitbucket.company.com".into()),
            source: SshKeySource::KeyContent {
                private_key: "secret".into(),
            },
            passphrase: Some("mypass".into()),
        };
        let json = serde_json::to_string(&config).unwrap();
        assert!(json.contains(r#""name":"Bitbucket Work""#));
        assert!(json.contains(r#""host":"bitbucket.company.com""#));
        assert!(json.contains(r#""kind":"keyContent""#));
        assert!(json.contains(r#""passphrase":"mypass""#));

        let parsed: SshKeyConfig = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed, config);
    }

    #[test]
    fn ssh_key_config_minimal_serialization() {
        let config = SshKeyConfig {
            name: "Simple Key".into(),
            host: None,
            source: SshKeySource::KeyFile {
                path: "~/.ssh/id_rsa".into(),
            },
            passphrase: None,
        };
        let json = serde_json::to_string(&config).unwrap();
        assert!(json.contains(r#""name":"Simple Key""#));
        assert!(json.contains(r#""host":null"#));
        assert!(json.contains(r#""passphrase":null"#));

        let parsed: SshKeyConfig = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed, config);
    }

    #[test]
    fn git_credentials_default_is_empty() {
        let creds = GitCredentials::default();
        assert!(creds.ssh_keys.is_empty());
    }

    #[test]
    fn git_credentials_serialization_empty() {
        let creds = GitCredentials::default();
        let json = serde_json::to_string(&creds).unwrap();
        assert_eq!(json, r#"{"sshKeys":[],"trustAllSshHostKeys":true}"#);

        // Deserializing the legacy format (without the field) defaults trust_all_ssh_host_keys to true.
        let parsed: GitCredentials = serde_json::from_str(r#"{"sshKeys":[]}"#).unwrap();
        assert!(parsed.ssh_keys.is_empty());
        assert!(parsed.trust_all_ssh_host_keys); // serde default = true
    }

    #[test]
    fn git_credentials_serialization_multiple_keys() {
        let creds = GitCredentials {
            ssh_keys: vec![
                SshKeyConfig {
                    name: "Key1".into(),
                    host: Some("github.com".into()),
                    source: SshKeySource::KeyContent {
                        private_key: "key1data".into(),
                    },
                    passphrase: None,
                },
                SshKeyConfig {
                    name: "Key2".into(),
                    host: None,
                    source: SshKeySource::KeyFile {
                        path: "/tmp/key".into(),
                    },
                    passphrase: Some("pw".into()),
                },
            ],
            trust_all_ssh_host_keys: true,
        };
        let json = serde_json::to_string(&creds).unwrap();
        assert!(json.contains("Key1"));
        assert!(json.contains("Key2"));
        assert!(json.contains("github.com"));
        assert!(json.contains("keyContent"));
        assert!(json.contains("keyFile"));

        let parsed: GitCredentials = serde_json::from_str(&json).unwrap();
        assert_eq!(parsed, creds);
    }

    #[test]
    fn git_credentials_deserialize_empty_json() {
        let parsed: GitCredentials = serde_json::from_str(r#"{}"#).unwrap();
        assert!(parsed.ssh_keys.is_empty());
    }

    #[test]
    fn git_credentials_deserialize_legacy_no_ssh_keys() {
        let parsed: GitCredentials = serde_json::from_str(r#"{"sshKeys":[]}"#).unwrap();
        assert!(parsed.ssh_keys.is_empty());
    }

    #[test]
    fn git_error_clone_failed_display() {
        let err = GitError::CloneFailed("test failure".into());
        assert_eq!(err.to_string(), "clone failed: test failure");
    }

    #[test]
    fn git_error_destination_exists_display() {
        let err = GitError::DestinationExists("/tmp/repo".into());
        assert_eq!(err.to_string(), "destination already exists: /tmp/repo");
    }
}
